HotelBase/HotelFactory.cs

```diff
@@ -1,27 +1,28 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using HotelBooking;
 
 namespace HotelBase
 {
     public class HotelFactory : IHotelFactory
     {
+        private readonly Dictionary<int, Type> _hotelTypesDictionary = new Dictionary<int, Type>
+            {{1, typeof(HotelExampleEmailCanFail)}, {2, typeof(HotelExample)}};
+
         public IHotel ReturnHotel(int id, IBookingService bookingService, IPaymentService paymentService,
             ILogger logger)
         {
-            switch (id)
+            if (_hotelTypesDictionary.TryGetValue(id, out var type))
             {
-                case 1:
-                    return new HotelExampleEmailCanFail(bookingService, paymentService, logger);
-                case 2:
-                    return new HotelExample(bookingService, paymentService, logger);
-                default:
-                    return null;
+                return (IHotel) Activator.CreateInstance(type, bookingService, paymentService, logger);
```

Comment:
Quite tricky, but much better than your switch implementation

Reply:
To be honest, I googled a lot and couldn't come up with any better working solution